Use schema at the time of the snapshot when reading a snapshot.

wypoon opened the pull request:
Iceberg uses the current schema of a table when reading an older snapshot of the table.
This change makes Iceberg use the schema of the table at the time of the snapshot instead.
To enable this, we add two new methods to the `Table` interface, `schemaForSnapshot` and `schemaForSnapshotAsOfTime`; the former takes the snapshot id and the latter takes a timestamp (millis since the epoch).
For Spark 2, we change `Reader` to use the appropriate schema if the snapshot-id or as-of-timestamp options are used.
For Spark 3, we change `SparkTable` as well as `SparkScanBuilder` and `SparkBatchScan` to use the appropriate schema.
We add two new unit tests that test the new behavior. They would fail without the above changes.

wypoon commented:
This is for https://github.com/apache/iceberg/issues/1501.

rdblue commented:
Thanks for working on this, @wypoon! This was just recently pointed out by another contributor as well. We're currently tracking this in #1029, which has a bit more detail about how this should be built.

While I like that your solution found a way to recover the old snapshot, we don't want to need to read old metadata files to recover it. Iceberg should track the schema that was used for each snapshot. That means a few steps are needed:

1. Add a list, `schemas`, to table metadata that has all of the schemas for known snapshots, like `partition-specs`
2. Add a `schema-id` to each schema so that the schemas can be referenced by id
3. Add a `schema-id` to each snapshot to track the schema that was current when it was created

Then we would need to add a `Snapshot.schema` method to retrieve the schema of a snapshot directly. Would you like to work on a PR for the first one?

wypoon commented:
@rdblue, thanks for the feedback and the pointer to #1029.
About step 1 in the steps you outlined, would the .metadata.json file then contain a `schema` field with the current schema and a (new) `schemas` field with a list of all the schemas (including the current)?
Sure, I'd be happy to work on a PR for this step.

rdblue commented:
Yes, we would add `schemas` to track recent schemas (referenced by any known snapshot) and keep `schema` for now. We would also add `current-schema-id` to track which schema in `schemas` is the current one. Eventually, we will deprecate `schema` and just use the one identified by id.

electrum commented:
Should we read the old metadata files as a fallback for data written before this addition?

rdblue commented:
> Should we read the old metadata files as a fallback for data written before this addition?

We could, but it doesn't seem worth the effort to maintain it to me. If other people disagree, I'm fine adding it.

electrum commented:
The question is what to do when we access a historical snapshot for older snapshots. It seems we have a few choices:

1. Use the current schema, as we do today.
2. Fail the query.
3. Fallback to reading old metadata files.

Using the current schema seems like the worst choice, as that results in inconsistent behavior for reasons that are invisible to the user. Failing the query would break behavior that works today.

While I agree that having more code to handle older data is annoying, the complexity here seems low (~20 lines of code) relative to the benefit. We have much higher complexity in other areas, such as supporting non-projected identity columns for converted Hive tables (and that's not even covered by the specification).

rdblue commented:
> While I agree that having more code to handle older data is annoying, the complexity here seems low (~20 lines of code) relative to the benefit.

I agree with that. And since we plan to make schema tracking required in v2, we can easily remove this work-around.

rdblue commented:
@wypoon, are you interested in working on both? We could update this PR to use the work-around you implemented and to update the `Snapshot` API (add `Snapshot.schema()`). Then we can add the new schema tracking separately. What do you think?

rdblue commented on api/src/main/java/org/apache/iceberg/Table.java:
```diff
@@ -51,6 +51,23 @@
+   *
+   * @return this table's schema
+   */
+  Schema schemaForSnapshotAsOfTime(long timestampMillis);
```
Instead of changing the table API in this commit, I think we only need to add a `schema` method to `Snapshot`. That will be used later when we have better tracking. Not adding too many methods to the public API in this commit will give us more flexibility later and make this simpler to get in.

wypoon replied:
I can avoid adding the `schemaForSnaphot` methods to the `Table` API. I think only `BaseTable` needs to have the methods. However, the places where I need to call `schemaForSnapshot` have references to simply `Table`; I'd have to check if the `Table` is an instance of `BaseTable`.
I also suggested in the general comments that `Snapshot` could simply have a `schemaId` method in the future. And that would then get used in a revised implementation of `schemaForSnapshot` in `BaseTable`.

rdblue replied:
Yeah, I agree with you. We have some flexibility here with how we look up the schema for a given snapshot. The main thing is that we don't want to expose any methods through the `Table` API right now. And since we don't have a schema id yet, we can't really add that method to `Snapshot`.

Eventually, I think we would want `Snapshot` to return its own schema rather than requiring a lookup, which is why I suggested that option. But we might be able to avoid the problem for now.

Like I said, I just don't think we want to add anything to `Table`.

wypoon replied:
I have removed the methods from `Table`, adding them only to `BaseTable`. Thus the `Table` API is unchanged.

rdblue commented on core/src/main/java/org/apache/iceberg/BaseTable.java:
```diff
@@ -59,6 +60,45 @@ public Schema schema() {
+  }
+
+  @Override
+  public Schema schemaForSnapshotAsOfTime(long timestampMillis) {
```
We don't need two methods. We can use the history log to find out what snapshot was current at a given time, then get the schema from that snapshot.

wypoon replied:
Isn't that what I'm doing here? This method can be considered a convenience method.

rdblue replied:
Yes, the concern is that we want to keep the `Table` API small and focused.

wypoon replied:
I have kept `schemaForSnapshotAsOfTime` for convenience. Since the `Table` API is unchanged, I think it is ok.

rdblue commented on spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -165,13 +165,23 @@
+        this.schema = SparkSchemaUtil.prune(getTableSchema(), requestedSchema, filterExpression(), caseSensitive);
       } else {
-        this.schema = table.schema();
+        this.schema = getTableSchema();
```
In addition to updates here, we will need to update schema handling in `BaseTableScan` so that the projection returned by the scan is based on the schema of the `Snapshot` that will be scanned.

wypoon replied:
This is the only point I don't understand. As far as I can tell, `BaseTableScan` is constructed with the user's requested schema. If the user is wanting to read from an old snapshot, then they should be supplying a requested schema compatible with the snapshot's schema. I couldn't see anything I needed to change in `BaseTableScan`. If I'm missing something here, can you please point out to me the parts I need to change?

I added a new test case where a requested schema is used, and that revealed a bug in the spark2 `IcebergSource#createReader`. In that method, there is a call to `SparkSchemaUtil.convert(Schema, StructType)` whose only purpose seems to be to check that the requested `StructType` does not contain fields in the `Schema` (for then the call would cause an exception), and here I needed to use the snapshot `Schema` rather than the table `Schema`. Other than that, the projection returned by the scan appears to be correct.

rdblue commented on spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -165,13 +165,23 @@
           TableProperties.PARQUET_BATCH_SIZE, TableProperties.PARQUET_BATCH_SIZE_DEFAULT));
   }
 
+  private Schema getTableSchema() {
```
I think this method name could be more descriptive. The verb "get" doesn't add anything so we don't normally use it. And this is also not the table schema, it is a snapshot's schema. How about renaming this to `snapshotSchema()`?

wypoon replied:
Renamed to `snapshotSchema` as you suggest.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -62,7 +62,7 @@ public SparkTable getTable(StructType schema, Transform[] partitioning, Map<Stri
 
     // Build Spark table based on Iceberg table, and return it
-    return new SparkTable(icebergTable, schema);
+    return new SparkTable(icebergTable, schema, options);
```
I think that `SparkTable` should include a `snapshotId` so that it returns the correct snapshot's schema to the analyzer, when possible. I don't think that it needs to include these options, when no other options are used by the table.

wypoon replied:
Ok, I can do that.

wypoon replied:
I extracted `snapshotId` and `asOfTimestamp` from `options` here if they appear and pass the values to `SparkTable` in its constructor instead of having `SparkTable` do that. I honestly think it's just half a dozen of one and six of the other.

edwinchoi commented:
I think you need to check `.snapshots()` not `.snapshotLog()`. The latter seems to only retain the history of changes since the table was created/replaced. You can see in `TableMetadata.buildReplacement` that the prior history entries are not retained in the replacement.

Also, from what I can tell, the metadata is always refreshed after the snapshot has been created (e.g., `SnapshotProducer.commit()`, `BaseTransaction.commitTransaction()`, etc... each refresh the metadata as part of the commit protocol). If that is always the case, then the loop should match the `Snapshot` with the first `MetadataLogEntry` where `MetadataLogEntry.timestampMillis() >= Snapshot.timestampMillis()`, and not the last `MetadataLogEntry.timestampMillis() <= Snapshot.timestampMillis()`.

As a sanity check, the code could check the expected snapshot-id against the `TableMetadata` that is read.

wypoon commented:
Thanks everyone for your interest in this work!
@rdblue, I think it makes sense to get a version of this PR in first, and separately, do the work to associate a schema with each snapshot, while aligning this PR with the other work as much as possible.
To check that we're on the same page, your original suggestion was to add a list `schemas` to table metadata, add a `schema-id` to each schema, and add a `schema-id` to each snapshot. When this is done, `Snapshot` should have a `schema()` method.
IIUC, the schema still has to be read from the table metadata file, referenced using the `schema-id` from the snapshot. Thus an implementation of `Snapshot` would need a reference to the table metadata, correct?
Right now, to get the schema, I need to start from the table metadata file. There is currently no obvious way to get from snapshot to table metadata. I'd need to change the implementation of `Snapshot` to have a reference to the current metadata. Does that make sense to you? Or am I missing something?

rdblue commented:
Good observations, @edwinchoi. I'm not sure why the snapshot log is cleared out when a table is replaced. That log should track the snapshot that was current a given time, which seems like the right one to use. We use it for time travel queries, for example. But since the log is not there, we can use the metadata file log instead as you suggested to find the snapshot that was current at some time.

We should not use the list of snapshots because there is no guarantee that any given snapshot was ever the current version (like a branch in git that is never the main branch).

rdblue commented:
> There is currently no obvious way to get from snapshot to table metadata

Yes, we will need to have a way to get the right schema from table metadata by ID. I think what might make sense for now is to pass a `Function<Long, Schema>` in that does the lookup by snapshot ID. Then the `schemaForSnapshot` method could be package-private and passed into the `Snapshot` through its constructor. All of this would be package-private so we can change it later as we need to.

wypoon commented:
> > There is currently no obvious way to get from snapshot to table metadata
> 
> Yes, we will need to have a way to get the right schema from table metadata by ID. I think what might make sense for now is to pass a `Function<Long, Schema>` in that does the lookup by snapshot ID. Then the `schemaForSnapshot` method could be package-private and passed into the `Snapshot` through its constructor. All of this would be package-private so we can change it later as we need to.

It seems to me that it'd be simpler if in the future, `Snapshot` would have a `schemaId()` method rather than a `schema()` method; since the places that need to obtain a schema for a `Snapshot` already have a reference to `TableMetadata` at hand, and only need to call `Snapshot.schemaId()` and then use it to look up the schema in the `TableMetadata`. So the implementation of `getSchemaForSnapshot(long)` in `BaseTable` can stay the way it is now, and change later when `Snapshot.schemaId()` becomes available.
What do you think?

rdblue commented:
I posted in a comment above, but I'll echo it here: I don't think we should be adding anything to the `Table` API that we are going to want to remove later. Since we don't have schema ids, I don't think we can add the API that we eventually want right now. Adding `Snapshot.schema` seems like something we can do earlier and that would be the most friendly API long term. That's why I suggest it. We can do something else, as long as we are careful and only add non-public methods.

edwinchoi commented:
@rdblue, how would a snapshot be recorded in the metadata if it weren't at some point current? Intermediate snapshots are removed on commit, so I would imagine the only case where the metadata could refer to snapshots that were never current is if the commit were rejected. In that case, the snapshots wouldn't be reachable from the current commit. What am I missing?

Example:
```sql
-- S1
CREATE TABLE test.ns.tbl USING iceberg AS
SELECT * FROM VALUES (1, "Alice"), (2, "Bob") AS (id, fname);

-- S2
CREATE OR REPLACE TABLE test.ns.tbl USING iceberg AS
SELECT * FROM VALUES (1, 5, "alice"), (2, 3, "bob") AS (id, name_len, name) ;

-- S3
INSERT INTO test.ns.tbl VALUES (3, 5, "carol")
```

The table `test.ns.tbl` has 3 valid states. So we should be able to read each one individually. `Table.history()`, after all 3 statements have executed, only returns entries for `S2` and `S3`; while `Table.snapshots()` returns all 3.

You can rewind to `S1`, so it seems time-travel is using snapshots not the log. Trying to read the data will fail since it tries to cast the 2nd field to an integer, but it will succeed if you rewrite the snippet to use `(id, name, name_len)`.

rdblue commented:
There are a few cases where a snapshot might not have been the current state:
1. For the write-audit-publish (WAP) pattern, there is an option to only stage a commit and not update the table's current-snapshot-id. In this case, the writer updates the table by creating a new snapshot. Then an auditor reads the snapshot and validates it (with row counts, for example), and if the snapshot looks good, commits the snapshot as the current table state. This allows reports to be validated before going live.
2. Single-table transactions will create individual snapshots, but the table state goes directly to the latest snapshot when the transaction commits.

Also, the table state can be rolled back to a previous snapshot and new commits will form a new history afterwards.

edwinchoi commented:
Thanks for elaborating. My thoughts...

> For the write-audit-publish (WAP) pattern, there is an option to only stage a commit and not update the table's current-snapshot-id. In this case, the writer updates the table by creating a new snapshot. Then an auditor reads the snapshot and validates it (with row counts, for example), and if the snapshot looks good, commits the snapshot as the current table state. This allows reports to be validated before going live.

I don't think WAP is working as expected under RTAS.

```scala
spark.sql("""
CREATE TABLE test.ns.tbl
USING iceberg
TBLPROPERTIES ('write.wap.enabled'='true')
AS SELECT * FROM VALUES (1, "Alice"), (2, "Bob") AS (id, fname)
""")
spark.conf.set("spark.wap.id", "12345")
spark.sql("""
CREATE OR REPLACE TABLE test.ns.tbl
USING iceberg
AS SELECT * FROM VALUES (1, 5, "alice"), (2, 3, "bob") AS (id, name_len, name)
""")
spark.conf.unset("spark.wap.id")
```

After running this, the schema from the staged change is showing up but the data that _should_ exist isn't accessible, i.e., `DESC test.ns.tbl` shows the new schema and `SELECT * FROM test.ns.tbl` is coming up empty. Even under a simple schema change `ALTER TABLE ... ADD COLUMN ...`, the change is taking effect prematurely.

> Also, the table state can be rolled back to a previous snapshot and new commits will form a new history afterwards.

I don't believe this changes the notion of what was current at some point in time. If you view the timeline as being from the database's point of view, then rolling back doesn't change the fact that at _some point in time_, a snapshot, that is now inaccessible, was visible in the database.

rdblue commented:
> I don't think WAP is working as expected under RTAS.

You're right. I think we should disable WAP with RTAS until we figure out what the behavior should be.

> I don't believe [a rollback] changes the notion of what was current at some point in time.

That's right, but it is an example of not being able to simply use the list of snapshots. This was a side point that supports the idea that point in time lookup should be done using the snapshot log, or metadata file log if it is not available, and snapshot lookup should be done strictly by id.

wypoon commented:
> I think you need to check `.snapshots()` not `.snapshotLog()`. The latter seems to only retain the history of changes since the table was created/replaced. You can see in `TableMetadata.buildReplacement` that the prior history entries are not retained in the replacement.
> 
> Also, from what I can tell, the metadata is always refreshed after the snapshot has been created (e.g., `SnapshotProducer.commit()`, `BaseTransaction.commitTransaction()`, etc... each refresh the metadata as part of the commit protocol). If that is always the case, then the loop should match the `Snapshot` with the first `MetadataLogEntry` where `MetadataLogEntry.timestampMillis() >= Snapshot.timestampMillis()`, and not the last `MetadataLogEntry.timestampMillis() <= Snapshot.timestampMillis()`.
> 
> As a sanity check, the code could check the expected snapshot-id against the `TableMetadata` that is read.

My understanding from @rdblue's replies is that one should use the snapshot log, so I have stuck with that. Also, from what I see, the metadata timestamp is always the same as the snapshot timestamp when the metadata is written for a new snapshot. I changed the loop to look for the metadata log entry whose timestamp == the snapshot timestamp (and break out of the loop). (I could also use >= as you suggest, but it should make no difference.)

edwinchoi commented:
> Also, from what I see, the metadata timestamp is always the same as the snapshot timestamp when the metadata is written for a new snapshot.

If you use Spark 3's catalog API, you'll see that the snapshot timestamp and the metadata are _not guaranteed_ to have the same time . You can trace the call from `SparkCatalog.stageCreateOrReplace`. RTAS applies the changes in a transaction, which uses two separate calls to `System.currentTimeMillis()`.

Try adding tests to `TestCreateTableAsSelect` that do CTAS/RTAS, and you'll see that the timestamps are not the same.

Also, after giving this some more thought, you can't rely on a partial ordering between the snapshot and metadata update timestamps. `System.currentTimeMillis()` is not monotonic - clock adjustments via NTP can cause two consecutive readings to go back in time. The only safe option then is to scan the metadata files to find the file where the current-snapshot-id matches the target snapshot-id.

edwinchoi commented:
@rdblue what about adding a method `Table rewind(Snapshot)` to `Table`? This seems like a more straight-forward way to handle views as-of a snapshot or moment in time with potentially enough flexibility to work with v2 (queries as-of a moment would select the snapshot upstream). The return value would be a new instance that prohibits any mutating operations against `Table` and `TableOperations`.

I suspect there would need to be an equivalent `rewind` on `TableOperations` as well for fetching the matching metadata file (there are a lot of fields in the metadata object that seem relevant to queries, e.g., partition specs, table properties, and sort orders.. I'm assuming it's necessary to know these to read a historic snapshot, and not simply the schema alone).

wypoon commented:
Sorry for the long pause.
My understanding is that the metadata-log (`TableMetadata.previousFiles()`) is in order of metadata committed, even if the timestamps may not be in order due to clock skew on different machines from which the commits are done. I have implemented @edwinchoi's suggestion to read each metadata file listed in the metadata-log to find the first one where the current-snapshot-id matches the target snapshot id.
I have rebased on master (probably behind again by now), and made some changes accordingly.
@rdblue, can you please take a look at this again?

waterlx commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -60,10 +61,13 @@ public SparkTable getTable(StructType schema, Transform[] partitioning, Map<Stri
     // Build Spark table based on Iceberg table, and return it
     // Eagerly refresh the table before reading to ensure views containing this table show up-to-date data
-    return new SparkTable(icebergTable, schema, true);
+    return new SparkTable(icebergTable, schema, snapshotId, asOfTimestamp, true);
```
@wypoon @rdblue I am trying to apply this patch to the most recently master branch, and find that the changes to this IcebergSource for Spark3 could not be applied due to #1783. "SparkTable" is "loadTable" by TableCatalog and in SparkCatalog, SparkTable is constructed by "new SparkTable(icebergTable, !cacheEnabled)" with snapshotId and asOfTimestamp set as null internally. It seems that options given by IcebergSource#getTable() could be not passed here. Could you please share your thoughts on that? Thanks!

wypoon replied:
@waterlx I rebased on master before #1783 was merged. #1783 does indeed cause difficulty here as you point out. I don't see a natural place to get the `options` passed into `IcebergSource#getTable` to where the `SparkTable` gets constructed in `SparkCatalog#loadTable`.
@rdblue do you have any suggestions?

wypoon replied:
Q. Is it the case that in master after #1783, there is no path to `new SparkTable(...)` where the  `requestedSchema` argument is not null?

Anyway, one idea that comes to my mind is that I could add some method to `SparkTable` to set the snapshotId and asOfTimestamp after an instance is already created, and in `IcebergSource#getTable`, once we have loaded a `Table` from the catalog, check if it is a `SparkTable` and then set the snapshotId and asOfTimestamp.

waterlx replied:
@wypoon it seems that in DataFrame load, the logic in IcebergSource (of Spark 3) #getTable() is not experienced.

waterlx replied:
@wypoon As exposed by the UT you added, when data frame read (`SparkCatalog#loadTable()` is called by `DataFrameReader`), I think we might need a Identifier with options here to bring snapshot-id and as-of-timestamp with the identifier. I tried with wrapping`org.apache.spark.sql.connector.catalog.Identifier` into a new interface and updating related code such as `Spark3Util` to construct an implentation of it in `catalogAndIdentifier` and it seems work here. But I think that is so hacked and rude.

BTW, in `testSnapshotReadAfterAddAndDropColumn`, I got `iceberg does not support user specified schema. Please don't specify the schema.` because a schema is specified when dataframe reader. I checked the Spark 3 code and found that error message is expected. But your UT seems not expect that. Could you please share your idea about that?

wypoon replied:
@waterlx, you are correct that when we invoke `DataFrameReader#load` in Spark 3, `IcebergSource#getTable` is not called; rather, `SparkCatalog#loadTable` is called directly. This is because #1783 changed the spark3 `IcebergSource` to be a `SupportsCatalogOptions`, not just a `TableProvider`. Therefore in `DataFrameReader#load`, this [logic](https://github.com/apache/spark/blob/2b147c4cd50da32fe2b4167f97c8142102a0510d/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L265-L271) is executed (which calls `SparkCatalog#loadTable`), where before, when `IcebergSource` was only a `TableProvider`, this [other logic](https://github.com/apache/spark/blob/2b147c4cd50da32fe2b4167f97c8142102a0510d/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L272-L275) was executed (which calls `IcebergSource#getTable`). (You already know this.)
Given this, in order for the `SparkTable` returned by `SparkCatalog#loadTable(Identifier)` to be aware of the snapshot, the information about the snapshot (such as is specified by `df.read.format("iceberg").option("snapshot-id", ...).load(...)`) needs to be present in the `Identifier`, as you concluded. I too don't see another way to do it. I have an implementation along this line that works. (At the moment, my implementation causes some failures in the spark3-extensions unit tests around relation cache refresh in snapshot-related procedures, so I need to figure that out.)
As for the failure that results in `iceberg does not support user specified schema`, I simply need to remove specifying a schema (`.schema("id INT, data STRING")`) when reading the Iceberg table, as `SupportsCatalogOptions` does not allow it. (This test was valid before #1783 but invalid after.) Related to this, in `SparkTable`, there is no use for a `requestedSchema` field anymore since we cannot specify a schema, so it should not be in the `SparkTable` constructor either. (I had posed this question above.)
In summary, I concur with your investigation and thank you for sharing your findings.

wypoon replied:
I fixed the cause of the spark3-extensions failures and updated this PR with my changes.

yyanyy commented:
> @rdblue, I think it makes sense to get a version of this PR in first, and separately, do the work to associate a schema with each snapshot, while aligning this PR with the other work as much as possible.

@wypoon  Sorry for the late jump in, since the schema-id change in table metadata that brought up earlier in this conversation thread is part of formatV2 release milestone tasks, I went ahead and started to implement part of it in #2096. Please take a look if you are interested, and please let me know if that change aligns with your plan. Thank you!

wypoon commented:
Retest please.

wypoon commented:
Tests passed for me locally. The
```
    java.lang.RuntimeException: Failed to get table info from metastore default.table

        Caused by:
        org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)

            Caused by:
            java.net.SocketException: Broken pipe (Write failed)

```
that caused failures seems to be a test problem unrelated to my changes.

wypoon commented:
@yyanyy thank you for your work in #2096 and #2275.
@rdblue can you please take a look at this again and let me know your thoughts on how to proceed with this? I think that there is still a place for this work. There are two parts to this work: 
1. `BaseTable#getSchemaForSnapshot`.
2. Changes to Spark 2 and Spark 3 to use the schema for the snapshot when reading. These directly or indirectly call `BaseTable#getSchemaForSnapshot`.
The first part can be changed to lookup the schema by `schema-id` from the `schemas` list now that snapshots in the `snapshot-log` have a `schema-id` with @yyanyy's work. The second part, I think, is still applicable.
One other aspect is that I'm not competent to make changes for the other supported components such as Hive and Flink, only for Spark, but the changes for those components could be done separately and by other folks.

wypoon commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkTable.java:
```diff
@@ -76,24 +77,21 @@
-      // convert the requested schema to throw an exception if any requested fields are unknown
-      SparkSchemaUtil.convert(icebergTable.schema(), requestedSchema);
-    }
   }
```
I removed `requestedSchema` from `SparkTable` because with #1783, the spark3 `IcebergSource` changed to be a `SupportsCatalogOptions`, not just a `TableProvider`. Since `DataFrameReader` does not support specifying a schema when reading from an `IcebergSource`:
```
    DataSource.lookupDataSourceV2(source, sparkSession.sessionState.conf).map { provider =>
      ...
      val (table, catalog, ident) = provider match {
        case _: SupportsCatalogOptions if userSpecifiedSchema.nonEmpty =>
          throw new IllegalArgumentException(
            s"$source does not support user specified schema. Please don't specify the schema.")
```
(see https://github.com/apache/spark/blob/v3.1.2/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L282-L285)
there is no reason to have a `requestedSchema` field as we cannot make use of it.

wypoon commented:
@yyanyy I have updated this since #2275 was merged. Can you please review it?
In `BaseTable#schemaForSnapshot`, we look up the schema by `schema-id` from the `schemas` list. If it is not found, we go through the previous metadata files until we find one whose current snapshot id is the snapshot id and get the schema from there.
The rest of the PR basically supports using the schema for the snapshot when reading an Iceberg table from Spark. I am not qualified to make the changes for other engines (I don't believe Hive supports time travel yet), so that can be done by others in separate PRs.
@rdblue, @RussellSpitzer, @aokolnychyi , can any of you help review this too?

rdblue commented on core/src/main/java/org/apache/iceberg/BaseTable.java:
```diff
@@ -74,6 +75,57 @@ public Schema schema() {
+   *
+   * @return the schema
+   */
+  public Schema schemaForSnapshot(long snapshotId) {
```
Now that snapshots can report the schema that was current when the snapshot was created, I think that the first PR should focus on using that schema and not recovering schemas from old metadata files. We may want to add this later, but it over-complicates getting the initial feature done.

wypoon replied:
I will remove this and the following method from `BaseTable`.
I will move the recovery of schema from previous metadata files (in the event that a snapshot does not have a schema id associated with it) to a separate PR.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/SparkUtil.java:
```diff
@@ -85,17 +85,17 @@ public static void validatePartitionTransforms(PartitionSpec spec) {
         // Assume the first element is a valid catalog
         String[] namespace = nameParts.subList(1, lastElementIndex).toArray(new String[0]);
-        return Pair.of(catalog, identiferProvider.apply(namespace, name));
+        return Pair.of(catalog, identifierProvider.apply(namespace, name));
```
This looks like an unrelated change. Can you move it to a separate PR? We want to keep this one small since it's quite large already.

wypoon replied:
I was calling into this code and took the opportunity to fix a typo. I'll move this typo fix into a separate PR.

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:
```diff
@@ -901,6 +915,199 @@ public void testPartitionsTable() {
+    List<Row> actualRecords3 = resultDf3.orderBy("id")
+        .collectAsList();
+
+    Assert.assertEquals("Records should match", expectedRecords, actualRecords3);
```
It would be nice to have an assertion that the schema also matches.

wypoon replied:
Done.

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:
```diff
@@ -901,6 +915,199 @@ public void testPartitionsTable() {
+
+    Assert.assertEquals("Records should match", expectedRecords, actualRecords);
+
+    long ts1 = System.currentTimeMillis();
```
This is likely to make the tests flaky because of different behavior with time. To avoid the problem, we added the method `SparkTestBase.waitUntilAfter` that accepts a current timestamp and waits until the timestamp returned by `currentTimeMillis` is strictly after the given timestamp. To fix this test, just change these lines to:

```java
long ts1 = waitUntilAfter(System.currentTimeMillis());
```

I also recommend using more readable names, like `beforeDropColumnMs`.

wypoon replied:
Thanks for the tip.

wypoon replied:
Also gave the timestamp more descriptive names.

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:
```diff
@@ -901,6 +915,199 @@ public void testPartitionsTable() {
+    List<Row> actualRecords2 = resultDf2.orderBy("id")
+        .collectAsList();
+
+    List<Row> expectedRecords2 = Lists.newArrayList(
```
It's hard to follow the `expectedRecords` variables in these tests. I think it would be better to give them descriptive names, like `originalRecords` or `expectedWithoutDroppedColumn`.

wypoon replied:
I renamed the `expectedRecords` variables as you suggested.

rdblue commented on spark2/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -68,8 +68,8 @@ public DataSourceReader createReader(StructType readSchema, DataSourceOptions op
-      // convert() will fail if readSchema contains fields not in table.schema()
-      SparkSchemaUtil.convert(table.schema(), readSchema);
+      // convert() will fail if readSchema contains fields not in reader.snapshotSchema()
+      SparkSchemaUtil.convert(reader.snapshotSchema(), readSchema);
```
I think it would be better to expose `reader.snapshotId()` and use the snapshot ID to lookup the snapshot and its schema. This shouldn't be needed in Spark 3 so I think it is fine if it's a bit uglier here.

rdblue replied:
Nevermind, I think it's okay the way it is working now.

rdblue commented on spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -152,20 +153,30 @@
   }
 
+  protected Schema snapshotSchema() {
+    if (snapshotId != null && table instanceof BaseTable) {
```
I think this should use the snapshot's schema ID, like this:

```java
if (snapshotId != null) {
  Schema snapshotSchema = table.schemas().get(table.snapshot(snapshotId).schemaId());
  if (snapshotSchema != null) {
    return snapshotSchema;
  }

  // TODO: recover the schema using a util class
  return table.schema();
} else {
  return table.schema();
}
```

wypoon replied:
Yes, I'll follow your suggestion so we don't need to change `BaseTable`.
Also, as you point out below, this same code appears in several places. I'll put it in a util method.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/Spark3Util.java:
```diff
@@ -742,16 +743,20 @@ private static String sqlString(org.apache.iceberg.expressions.Literal<?> lit) {
     return null;
   }
 
+  private static List<String> getMultipartIdentifier(SparkSession spark, String name) throws ParseException {
```
Nit: use of `get` in a method name signals either that there is a more descriptive verb (like "create") or that "get" can simply be removed (as in the case of a getter).

wypoon replied:
Will change.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/Spark3Util.java:
```diff
@@ -767,19 +772,46 @@ public static CatalogAndIdentifier catalogAndIdentifier(String description, Spar
     }
   }
 
+  public static CatalogAndIdentifier catalogAndIdentifier(String description, SparkSession spark, String name,
```
Rather than making this resolution more complicated with cases for timestamp and snapshot ID, can we make these orthogonal? What about calling `catalogAndIdentifier` and then adding the snapshot to the path identifier?

wypoon replied:
I tried leaving the `catalogAndIdentifier` methods in `Spark3Util` unchanged, and adding the snapshot id and timestamp to the Identifier in (spark3) `IcebergSource#catalogAndIdentifier` after calling `Spark3Util.catalogAndIdentifier`, but I must have missed something because the unit tests I added then failed. So I reverted that experiment. This implementation may be slightly more complicated but I know that it works.

wypoon replied:
I figured out what I missed the first time around. I now add no new methods to `Spark3Util` and use existing `Spark3Util.catalogAndIdentifier` methods to obtain a `CatalogAndIdentifier`, and then swap out the `Identifier` for a snapshot-aware one.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/SnapshotAwareIdentifier.java:
```diff
@@ -0,0 +1,36 @@
+
+  Long snapshotId();
+
+  Long asOfTimestamp();
```
I'm skeptical that it's a good idea to pass both snapshot ID and the timestamp. Could the timestamp be easily resolved to a snapshot before creating the identifier so that we have one that is easier to work with?

wypoon replied:
The problem is that we need the `Table` in order to resolve the timestamp to a snapshot, and we need the identifier in order to load the `Table`. (Once the `Table` is loaded, we don't have an opportunity to set the timestamp, hence the need for `SnapshotAwareIdentifier`.) That is why I put the snapshot id and timestamp (if either (or both) appear) in the identifier, and deal with resolving to a snapshot later.
Incidentally, I raised https://github.com/apache/iceberg/issues/3070 in case both snapshot id and timestamp are specified.

rdblue replied:
Ah, understood. Thank you for explaining.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkBatchQueryScan.java:
```diff
@@ -79,6 +80,17 @@
   }
 
+  @Override
+  protected Schema snapshotSchema() {
```
Why is this method here? The scan should be passed a schema because all of the projection happens in the builder.

wypoon replied:
After working through how Spark constructs a `SparkBatchQueryScan`, I realize that you're right and that this method is not needed in `SparkBatchScan`.

wypoon replied:
Removed `snapshotSchema()` from SparkBatchScan and its subclasses.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:
```diff
@@ -64,16 +69,28 @@
+      return ((BaseTable) table).schemaForSnapshotAsOfTime(asOfTimestamp);
+    } else {
+      return table.schema();
+    }
```
Same as above, I think this should use the snapshot's schema or the table schema for now. Recovery can happen in a later update.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkTable.java:
```diff
@@ -113,14 +111,20 @@ public String name() {
+      return ((BaseTable) icebergTable).schemaForSnapshotAsOfTime(asOfTimestamp);
+    } else {
+      return icebergTable.schema();
+    }
```
There are 4 copies of this logic. I think it would be good to put this in a util class, like `SnapshotUtil`. Maybe name the method `SnapshotUtil.schemaAsOf(Table, Snapshot)`

rdblue reviewed (changes requested):
Thanks @wypoon! This is looking close to being ready.

One more question beyond the comments I just made: why doesn't this update `BaseTableScan`? Shouldn't the scan also use the schema from the correct snapshot? Would that help simplify what is happening here?

wypoon commented:
@rdblue, thanks for your detailed review. I'm working through addressing your feedback and will post an update very soon. I also looked more carefully at `BaseTableScan` and will update the schema there when `useSnapshot` is called.

wypoon commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkBatchScan.java:
```diff
@@ -207,9 +207,8 @@ public Statistics estimateStatistics() {
       LOG.debug("using table metadata to estimate table statistics");
       long totalRecords = PropertyUtil.propertyAsLong(table.currentSnapshot().summary(),
           SnapshotSummary.TOTAL_RECORDS_PROP, Long.MAX_VALUE);
-      Schema projectedSchema = expectedSchema != null ? expectedSchema : table.schema();
```
`expectedSchema` should not be null.
(Previously, I was replacing calls to `table.schema()` with calls to a method that returns the snapshot schema, but it should not be necessary in `SparkBatchScan`.)

wypoon commented:
@rdblue, I have rebased on master and updated the PR with most of your suggestions. The only change you suggested that I did not adopt concerns having both `snapshotId` and `asOfTimestamp` in `SnapshotAwareIdentifier`. I explained my reason above.
I hope that this is ready to merge now.

rdblue commented on core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:
```diff
@@ -144,4 +146,72 @@ public static Snapshot snapshotAfter(Table table, long snapshotId) {
+   * Returns the ID of the most recent snapshot for the table as of the timestamp.
+   *
+   * @param table a {@link Table}
+   * @param timestampMillis the timestamp in millis since the epoch
```
When mentioning epoch, I think it should state _which_ epoch. I'd update this to "Unix epoch"

wypoon replied:
Ok

rdblue commented on spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -166,13 +167,17 @@
   }
 
+  protected Schema snapshotSchema() {
+    return SnapshotUtil.schemaFor(table, snapshotId, asOfTimestamp);
```
I don't think that this should determine the snapshot that will be scanned in multiple places, but right now the logic to find the snapshot is here and in `tasks()`.

Instead, I think this class should create a `TableScan`, configure it using the snapshot selection criteria, and use the `schema()` from the table scan here. In `tasks()`, this base scan can be further refined before getting tasks (each `TableScan` object is immutable and independent).

rdblue replied:
I have a working version of this. I'll post a PR against your branch.

rdblue replied:
https://github.com/wypoon/iceberg/pull/1

wypoon replied:
Incorporated refactored version of wypoon#1.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/TableIdentifier.java:
```diff
@@ -0,0 +1,60 @@
+
+import org.apache.spark.sql.connector.catalog.Identifier;
+
+public class TableIdentifier implements SnapshotAwareIdentifier {
```
There is a `TableIdentifier` in `org.apache.iceberg.catalog` already. Can we make this a `SparkTableIdentifier` instead?

wypoon replied:
Yes, I agree that it is confusing to have two `TableIdentifier`s in Iceberg.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -101,24 +103,35 @@ public Table getTable(StructType schema, Transform[] partitioning, Map<String, S
       return new Spark3Util.CatalogAndIdentifier(catalogManager.catalog(DEFAULT_CATALOG_NAME),
-          catalogAndIdentifier.identifier());
-    } else {
+          TableIdentifier.of(ident.namespace(), ident.name(), snapshotId, asOfTimestamp));
```
Is this necessary? I thought that the scan would still be constructed in Spark3 with the `options` passed here. Are these options independent of read options?

wypoon replied:
I'm not sure what you're asking.
Here is where I implement your suggestion from your previous review, which was, instead of adding more `Spark3Util.catalogAndIdentifier` methods to handle `snapshotId` and `asOfTimestamp` parameters, simply to resolve the `CatalogAndIdentifier` and add `snapshotId` and `asOfTimestamp` to the `Identifier`. I wanted to keep the `TableIdentifier` immutable (not have methods to set `snapshotId` and `asOfTimestamp` after it has been constructed), so here I simply swap out the `Identifier`. This (and in the final else clause) is where it happens.

wypoon replied:
I don't recall the exact pathway, but the unit tests actually use a `Catalog` that is named "spark_catalog" but is not an instance of `SparkSessionCatalog`, so this has to be done here, or the unit tests will fail.

rdblue replied:
Okay, thanks.

rdblue commented on spark3/src/test/java/org/apache/iceberg/spark/source/TestPathIdentifier.java:
```diff
@@ -59,7 +59,7 @@
   public void before() throws IOException {
     tableLocation = temp.newFolder();
-    identifier = new PathIdentifier(tableLocation.getAbsolutePath());
+    identifier = new PathIdentifier(tableLocation.getAbsolutePath(), null, null);
```
Rather than changing this file, can you add a constructor to fill these in as defaults?

wypoon replied:
Ok

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:
```diff
@@ -1112,6 +1126,239 @@ public void testPartitionsTable() {
+        RowFactory.create(3, "z", null),
+        RowFactory.create(4, "xy", "B"),
+        RowFactory.create(5, "xyz", "C"));
+    Assert.assertEquals("Records should match", updatedRecords, actualRecords2);
```
Rather than using names like `actualRecords2`, you can just move the `collectAsList` into the assertion.

rdblue replied:
I should mention that these tests look good otherwise.

wypoon replied:
Adopted your suggestion.

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestIcebergSourceTablesBase.java:
```diff
@@ -1112,6 +1126,239 @@ public void testPartitionsTable() {
+        .collectAsList();
+
+    // At tsAfterDropColumn, there has been a schema change, but no new snapshot,
+    // so the snapshot as of tsAfterDropColumn is the same as that as of tsBeforeDropColumn.
```
This makes me think that we should only time travel with schema if `snapshot-id` is set and not `as-of-timestamp` because this behavior is inaccurate. If you read the table at time `tsAfterDropColumn`, then the result would be different. That said, I think this is a consequence of a very reasonable implementation. `as-of-timestamp` selects a snapshot and we read with the snapshot's schema.

@RussellSpitzer and @yyanyy, what do you think?

rdblue replied:
@wypoon, I don't think this should block this PR. This PR looks good and we can figure out what the behavior of `as-of-timestamp` should be separately.

wypoon replied:
Thanks. As you say, the behavior of `as-of-timestamp` is a consequence of a very reasonable implementation. Currently, we associate a `schema-id` with a snapshot, but there is no new snapshot written when just the schema is changed.

rdblue commented on core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:
```diff
@@ -144,4 +146,72 @@ public static Snapshot snapshotAfter(Table table, long snapshotId) {
+    Integer schemaId = snapshot.schemaId();
+    // schemaId could be null, if snapshot was created before Iceberg added schema id to snapshot
+    if (schemaId != null) {
+      return table.schemas().get(schemaId);
```
How about `getOrDefault(schemaId, table.schema())` just to be safe?

CodingCat replied:
shall we allow the current table.schema as the default? IMO, the current behavior is actually confusing, especially when the schema is not compatible, for example 

first snapshot
```
+-----+
|value|
+-----+
|    1|
|    2|
|    3|
|    4|
+-----+
```
overwrite with

```
+---------+
|new_value|
+---------+
|        2|
|        3|
|        4|
|        5|
+---------+
```

time travel to the first one

```
+---------+
|new_value|
+---------+
|        0|
|        0|
|        0|
|        0|
+---------+

```

CodingCat replied:
or we should add some configuration entry for users to explicitly allow time traveling to old data but use the latest schema? (IMHO, better than falling into the trouble as I described above)

wypoon replied:
IMO, we should not silently return `table.schema()` in this case; i.e., I don't think we should use `getOrDefault(schemaId, table.schema())`. This is the case where the snapshot does have a `schemaId`, so the schema should be found in `schemas()`. If it isn't, there was some internal inconsistency or corrupt metadata.

We only fall back to returning `table.schema()` as an interim measure (indicated by the TODO) in the following, until the recovery is implemented.

rdblue replied:
@wypoon, that's a fair point. In that case, we should make sure we throw a good error message. Probably use `IllegalStateException`?

wypoon replied:
Yes, I'll throw an `IllegalStateException` if the schema is not found then.

rdblue commented on core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:
```diff
@@ -144,4 +146,72 @@ public static Snapshot snapshotAfter(Table table, long snapshotId) {
+        snapshotId = logEntry.snapshotId();
+      }
+    }
+    Preconditions.checkArgument(snapshotId != null,
```
Nit: could you add a newline between control flow blocks and the following statement?

wypoon replied:
Added a blank line after the block.

wypoon commented:
@rdblue, when you have the time, this is ready for another review. I believe I have incorporated all your feedback. I have incorporated your proposed change (after some refactoring) to the spark `Reader` to configure a base `TableScan` on construction, and let the `TableScan` get the schema for the snapshot; as well as the other suggestions. I also fixed a bug with my last change to `BaseTableScan#useSnapshot`.

CodingCat commented:
@rdblue is this ready to merge? we really like the capability for iceberg to travel to the correct schema

rdblue commented on core/src/main/java/org/apache/iceberg/BaseTableScan.java:
```diff
@@ -123,8 +124,14 @@ public TableScan useSnapshot(long scanSnapshotId) {
         "Cannot find snapshot with ID %s", scanSnapshotId);
-    return newRefinedScan(
-        ops, table, schema, context.useSnapshotId(scanSnapshotId));
+    if (this instanceof DataTableScan) {
```
Instead of checking for a specific subclass, I think it is cleaner to override the method in `DataTableScan`. Here's what I was able to get working:

```java
  @Override
  public TableScan useSnapshot(long scanSnapshotId) {
    TableScan newScan = super.useSnapshot(scanSnapshotId);

    if (newScan.snapshot().schemaId() != null) {
      Schema snapshotSchema = SnapshotUtil.schemaFor(table(), scanSnapshotId);
      return newRefinedScan(tableOps(), table(), snapshotSchema, context().useSnapshotId(scanSnapshotId));
    }

    return newScan;
  }
```

wypoon replied:
Ok, I'll override the method in `DataTableScan`.

wypoon replied:
Q. (In your example) why check if `newScan.snapshot().schemaId() != null`? We can simply call `SnapshotUtil.schemaFor(table(), scanSnapshotId)` without checking, since the util method currently returns the table schema as a fallback, pending a TODO to recover the schema for snapshot by reading old metadata. This way, once that TODO is implemented, we don't need to revisit this code.
We don't need `newScan` at all; the only reason for it would be to call `super.useSnapshot` for its side-effects (argument validation), which I prefer to avoid.

rdblue replied:
The purpose of calling `newScan` was to use the argument validation in `BaseTableScan`. I'd rather not update that and not know that it needs to be updated here as well.

rdblue commented on core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:
```diff
@@ -144,4 +146,76 @@ public static Snapshot snapshotAfter(Table table, long snapshotId) {
+   */
+  public static Schema schemaFor(Table table, Long snapshotId, Long timestampMillis) {
+    Preconditions.checkArgument(snapshotId == null || timestampMillis == null,
+        "snapshot id and timestamp cannot both be used to find a schema");
```
Style: please capitalize the first word in error messages.

rdblue commented on spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -157,22 +128,88 @@
     this.readTimestampWithoutZone = SparkUtil.canHandleTimestampWithoutZone(options.asMap(), sessionConf);
   }
 
+  private void validateOptions(
```
It looks like the purpose of this method is to avoid updating tests because error messages are slightly different. Is that right? I'd prefer updating the tests so that we don't continue to duplicate checks.

wypoon replied:
There are two checks here. The first check does not seem to be tested by any existing unit test. The second check is tested by an existing unit test. Both checks also appear in the spark3 `SparkBatchQueryScan`. For that reason, I am preserving them here. I commented in https://github.com/wypoon/iceberg/pull/1 that I'd like to defer the clean up of duplicate checks to a separate PR, as it should be done for both spark2 and spark3 (and here we're only refactoring the spark2 `Reader`). Can that cleanup be done after this is merged, since as you remarked before, the change is already quite large?

rdblue replied:
I guess so, since this is already done.

rdblue commented on core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:
```diff
@@ -144,4 +146,76 @@ public static Snapshot snapshotAfter(Table table, long snapshotId) {
+
+    Preconditions.checkArgument(snapshotId != null,
+        "Cannot find a snapshot older than %s", timestampMillis);
+    return snapshotId;
```
Could you preserve more of the original method that was copied? For example, this no longer formats the timestamp in the error message, which makes it harder to interpret for users.

wypoon replied:
Ok, will do.

rdblue commented on core/src/main/java/org/apache/iceberg/BaseTableScan.java:
```diff
@@ -123,8 +124,14 @@ public TableScan useSnapshot(long scanSnapshotId) {
+          ops, table, snapshotSchema, context.useSnapshotId(scanSnapshotId));
+    } else {
+      return newRefinedScan(
+          ops, table, schema, context.useSnapshotId(scanSnapshotId));
```
Looks like the implementation of `SnapshotUtil.snapshotIdAsOfTime` was copied from the `asOfTime` method below, but the method below wasn't updated to use the utility. Can you fix that? I think you can change it to this:

```java
  @Override
  public TableScan asOfTime(long timestampMillis) {
    Preconditions.checkArgument(context.snapshotId() == null,
        "Cannot override snapshot, already set to id=%s", context.snapshotId());

    return useSnapshot(SnapshotUtil.snapshotIdAsOfTime(table(), timestampMillis));
  }
```

wypoon replied:
You're right. I'll have `BaseTableScan#asOfTime` call `SnapshotUtil.snapshotIdAsOfTime` to avoid duplicating the logic.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java:
```diff
@@ -101,24 +103,35 @@ public Table getTable(StructType schema, Transform[] partitioning, Map<String, S
     setupDefaultSparkCatalog(spark);
     String path = options.get("path");
+    Long snapshotId = Spark3Util.propertyAsLong(options, SparkReadOptions.SNAPSHOT_ID, null);
+    Long asOfTimestamp = Spark3Util.propertyAsLong(options, SparkReadOptions.AS_OF_TIMESTAMP, null);
```
I think that modifications to the source for snapshot and timestamp are unnecessary. Looking at the Spark code, the table will be loaded using the [catalog/identifier from `SupportsCatalogOptions`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala#L114-L120) and then [wrapped in a `DataSourceV2Relation` that uses the reader options](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala#L131). Those read options still include `snapshot-id` and `as-of-timestamp`.

When a scan is created, those [options get passed into the scan builder](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala#L42-L45).

There should be no need to handle these in `IcebergSource` because the options will be set at planning time. Not considering them here avoids the need to add `SnapshotAwareIdentifier` and removes ambiguous cases where the scan options identify a snapshot and the identifier did as well.

Can you update to remove this?

wypoon replied:
Let me look into this.

wypoon replied:
It turns out that you are mistaken.
The [`DataSourceV2Relation`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala#L131) is [created](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala#L175-L176) with the output attributes from the schema given by `SparkTable#schema()`.
The `SparkTable` is loaded using the catalog and identifier, and that is why I need the `SnapshotAwareIdentifier` when loading it, so that I can return the snapshot schema in `SparkTable#schema()`.
Otherwise the `SparkScanBuilder` does have the options for `snapshot-id` or `as-of-timestamp`, but its `pruneColumns()` will be called by Spark with a `requestedSchema` that is a subset of the table schema. Then its `build()` will return a `SparkBatchQueryScan` with an incorrect schema.
Once I remove the modifications to `IcebergSource`, `SparkCatalog` and `SparkTable`, then the unit tests I added all fail, as I suspected.

rdblue replied:
Great, thanks for looking into it more! I've been considering other options to solve this problem and I think we probably will need to go with something similar to what you have here.

The output attributes problem that you found makes sense. If there are columns in the snapshot's schema that have been dropped, then Spark doesn't have attributes for those columns and can't project them. We could try to expose the deleted columns as metadata columns in 3.1 and later, but we would still have a problem with `*` expansion because that uses the initial set of columns reported by the table and not metadata columns. So `select *` wouldn't behave the same way when time traveling if we took the approach with metadata columns.

We could also try to solve this on the Spark side, but I don't think that it's unreasonable for Spark to do what it is here and expect that the table reports a schema initially that can be used for query analysis.

The option that's left is loading the correct snapshot of the table so I agree with this direction. I'm weighing whether this should be done like metadata tables in the Iceberg layer or above it in Spark like you have here.

rdblue commented:
@wypoon, since #3269 is under discussion and it may take a while for us to decide how to pass the information for Spark 3, how about removing those portions here and getting the rest of this commit in?

wypoon commented:
Rebased on master. This was more painful than expected.
I'll proceed now with removal of the Spark 3 part of the change.

wypoon commented:
@rdblue I have rebased on master, and removed the spark3 part of the change. I also updated the PR description.
